server/src/gabriel_server/zeromq_server.py

```diff
@@ -0,0 +1,301 @@
+        self._is_running = True
+        asyncio.create_task(self._handler())
+        self._start_event.set()
+        logger.info(f"Listening on {URI_FORMAT.format(port)}")
+
+    async def wait_for_start(self):
+        await self._start_event.wait()
+
+    async def send_result_wrapper(
+            self, address, source_name, frame_id, result_wrapper, return_token):
+        """
+        Send result to client at address.
+
+        Args:
+            address: The identifier of the client to send the result to
+            source_name: The name of the source that the result corresponds to
+            frame_id: The frame id of the input that the result corresponds to
+            result_wrapper: The result payload to send to the client
+            return_token: Whether to return a token to the client
+
+        Returns True if send succeeded.
+        """
+        client = self._clients.get(address)
+        if client is None:
+            logger.warning('Send request to invalid address: %s', address)
+            return False
+
+        if source_name not in client.tokens_for_source:
+            logger.warning('Send request with invalid source: %s', source_name)
+            # Still send so client gets back token
+        elif return_token:
+            client.tokens_for_source[source_name] += 1
+
+        to_client = gabriel_pb2.ToClient()
+        to_client.response.source_name = source_name
+        to_client.response.frame_id = frame_id
+        to_client.response.return_token = return_token
+        to_client.response.result_wrapper.CopyFrom(result_wrapper)
+
+        logger.debug('Sending result to client %s', address)
+        await socket.send_multipart([
+            address,
+            to_client.SerializeToString()
+        ])
+
+    def add_source_consumed(self, source_name):
+        """
+        Indicate that at least one cognitive engine consumes frames from
+        source_name.
+
+        Args:
+            source_name (str): The name of the source to add
+
+        Must be called before self.launch() or run on the same event loop that
+        self.launch() uses.
+        """
+
+        if source_name in self._sources_consumed:
+            return
+
+        self._sources_consumed.add(source_name)
+        for client in self._clients.values():
+            client.tokens_for_source[source_name] = self._num_tokens_per_source
+            # TODO inform client about new source
+
+    def remove_source_consumed(self, source_name):
+        """
+        Indicate that all cognitive engines that consumed frames from source
+        have stopped.
+
+        Args:
+            source_name (str): The name of the source to remove
+
+        Must be called before self.launch() or run on the same event loop that
+        self.launch() uses.
+        """
+        if source_name not in self._sources_consumed:
+            return
+
+        self._sources_consumed.remove(source_name)
+        for client in self._clients.values():
```

Comment:
possible race condition? Or at least deleting from a list or dictionary in a loop needs to make sure ordering doesn't change, i think you need to do `for client in list(self._clients.values()):` to avoid the generator/reordering of list values.

Reply:
I'm not sure I follow. We are iterating through the clients, yes, but deleting an entry from a dictionary field of each client separately. We are not deleting any clients. Please correct me if I am misunderstanding your concern.

Can the ordering of the loop change just because I modify a value of the named tuple dictionary that I am iterating through? If we consider a simple case, a dictionary with integer values, will adding one to each value be wrong?